src/QueueBuffer.mo

```diff
@@ -194,26 +194,47 @@ module {
     public func len() : Nat = size() - deleted();
   };
 
+  /// A queue with fast random access, which preserves history of popped values
   public class QueueBuffer<X>() {
     var buf : Buffer<X> = Buffer<X>();
     var head : Nat = 0;
 
+    /// get id of oldest item in queue history
+    public func histId() : Nat = buf.deleted();
+    /// get id of oldest item in queue
+    public func headId() : Nat = head;
+    /// get next id which will be issued
+    public func nextId() : Nat = buf.size();
```

Comment:
About the naming: Id is mostly used for "identity", not "index" (which sometimes is "idx"). Here, at the library level, there are no semantics like identity and we should call these indices like we already do for example in the `get(index : Nat)` signature. 

Maybe we can spell this out as in `headIndex`, etc.?

Re: `nextIndex`. This cannot be named `tailIndex` for consistency, or? I believe nextIndex is probably tailIndex + 1. If so, then yes, we need a different name than tailIndex, indeed.

Regarding `histIndex`: Since we have the rewind function that implicitly uses it, shall we call it `rewindIndex`?

Reply:
headIndex returns the same Nat that is returned by peek, isn't it? Do we need headIndex then?

Reply:
Maybe headIndex can become peekIndex because it does the same but only returns the index.

Reply:
Agree with naming. Yes, nextIndex == tailIndex + 1, that's why I decided to use another name, to avoid confusion. Next index is more convenient to use, because in case of empty buffer the tailIndex cannot be simple natural number, and nextIndex is always valid

Reply:
That's right. But we have the same problem for headIndex and historyIndex which could also be "invalid" if the queue is empty and if they are intended as being inclusive.

I has also previously thought of pushCtr, popCtr, pruneCtr which makes sense even if the values are 0.

Reply:
regarding headIndex, it's not always true. `headIndex` can be used for reading amount of popped values, in our particular case to determine whether index is too old or too new. `peek` in case of empty queue will return null, so we cannot obtain index in this case. I think we'd better have a separate function

Reply:
Ok, so in the end I would say keep the names but use the longer versions (headIndex, historyIndex, nextIndex)

Reply:
Done